File: code/code64.py
from typing import List
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        n = len(grid)
        m = len(grid[0])
        dp = [[0] * m for _ in range(n)]
        dp[0][0] = grid[0][0]
        for index in range(1,n):
            dp[index][0] = dp[index-1][0] + grid[index][0]
        for index in range(1,m):
            dp[0][index] = dp[0][index-1] + grid[0][index]
        for i in range(1,n):
            for j in range(1,m):
                dp[i][j] = min(dp[i-1][j],dp[i][j-1]) + grid[i][j] 
        return dp[n-1][m-1]
if __name__ == "__main__":
    solution = Solution()
    grid = [[1,3,1],[1,5,1],[4,2,1]]
    ans = solution.minPathSum(grid)
    print(ans)

File: code/code64.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int n = grid.size(),m = grid[0].size();
        vector<vector<int>> dp(n,vector<int>(m,0));
        dp[0][0] = grid[0][0];
        for (auto i = 0;i < n;i++) {
            for (auto j = 0;j < m;j++) {
                int dx = i - 1,dy = j - 1;
                if (dx >= 0) dp[i][j] = dp[dx][j]+grid[i][j];
                if (dy >= 0) dp[i][j] = dp[i][dy]+grid[i][j];
                if (dx >= 0&&dy >= 0) dp[i][j] = min(dp[dx][j],dp[i][dy]) + grid[i][j];
            }
        }
        return dp[n-1][m-1];
    }
};
int main() {
    vector<vector<int>> grid = {{1,3,1},{1,5,1},{4,2,1}};
    Solution solution;
    int ans = solution.minPathSum(grid);
    cout << ans << endl;
}